src/blas/impl/KokkosBlas3_gemm_spec.hpp

```diff
@@ -135,73 +136,98 @@ struct GEMM {
-    gemm.run(team_size,vector_length,scratch_level);
-  }
-  if((transA[0]=='N' || transA[0]=='n') && (transB[0]=='T' || transB[0]=='t')) {
-    KokkosBlas::Impl::GEMMImpl<typename CViewType::execution_space,AViewType,BViewType,CViewType,blockA0,blockA1,blockB1,0,1> gemm(alpha,A,B,beta,C);
-    gemm.run(team_size,vector_length,scratch_level);
-  }
-  if((transA[0]=='T' || transA[0]=='t') && (transB[0]=='T' || transB[0]=='t')) {
-    KokkosBlas::Impl::GEMMImpl<typename CViewType::execution_space,AViewType,BViewType,CViewType,blockA0,blockA1,blockB1,1,1> gemm(alpha,A,B,beta,C);
-    gemm.run(team_size,vector_length,scratch_level);
-  }
-  if((transA[0]=='C' || transA[0]=='c') && (transB[0]=='T' || transB[0]=='t')) {
-    KokkosBlas::Impl::GEMMImpl<typename CViewType::execution_space,AViewType,BViewType,CViewType,blockA0,blockA1,blockB1,2,1> gemm(alpha,A,B,beta,C);
-    gemm.run(team_size,vector_length,scratch_level);
-  }
-  if((transA[0]=='N' || transA[0]=='n') && (transB[0]=='C' || transB[0]=='c')) {
-    KokkosBlas::Impl::GEMMImpl<typename CViewType::execution_space,AViewType,BViewType,CViewType,blockA0,blockA1,blockB1,0,2> gemm(alpha,A,B,beta,C);
-    gemm.run(team_size,vector_length,scratch_level);
-  }
-  if((transA[0]=='T' || transA[0]=='t') && (transB[0]=='C' || transB[0]=='c')) {
-    KokkosBlas::Impl::GEMMImpl<typename CViewType::execution_space,AViewType,BViewType,CViewType,blockA0,blockA1,blockB1,1,2> gemm(alpha,A,B,beta,C);
-    gemm.run(team_size,vector_length,scratch_level);
-  }
-  if((transA[0]=='C' || transA[0]=='c') && (transB[0]=='C' || transB[0]=='c')) {
-    KokkosBlas::Impl::GEMMImpl<typename CViewType::execution_space,AViewType,BViewType,CViewType,blockA0,blockA1,blockB1,2,2> gemm(alpha,A,B,beta,C);
-    gemm.run(team_size,vector_length,scratch_level);
+
+  typedef typename CViewType::execution_space ExecSpace;
+
+  // Figure out whether to use DotBased implementation
+  const int M = static_cast<int> (C.extent(0));
+  const int N = static_cast<int> (C.extent(1));
+
+  const bool A_is_lr = std::is_same<Kokkos::LayoutRight, typename AViewType::array_layout>::value;
+  const bool A_is_tr = ((transA[0]=='T') || (transA[0]=='t') || (transA[0]=='C') || (transA[0]=='c'));
+  const bool B_is_tr = ((transB[0]=='T') || (transB[0]=='t') || (transB[0]=='C') || (transB[0]=='c'));
+
+  // NOTE: these thresholds were copied from TPL CUBLAS, and may need to be retuned
+  constexpr int numDotsLayoutLeftThreshold = 1600;
+  constexpr int numDotsLayoutRightThreshold = 100;
+  if(    (!A_is_lr && A_is_tr && !B_is_tr && M*N < numDotsLayoutLeftThreshold)
+      || ( A_is_lr && A_is_tr && !B_is_tr && M*N < numDotsLayoutRightThreshold)) {
+
+    // call dot-based GEMM, only for C := beta * C + alpha * A^T * B
+    bool A_is_conj = ((transA[0]=='C') || (transA[0]=='c'));
+    DotBasedGEMM<ExecSpace, AViewType, BViewType, CViewType> dotBasedGemm(alpha, A, B, beta, C);
+    dotBasedGemm.run(A_is_conj ? true : false);
```

Comment:
I am pretty sure that the tertiary operator can be replaced by `A_is_conj` itself since it is already a boolean.